File: ch02/prime3.c
#include<stdio.h>

#define N 1000000

int main(void)
{
		int i, n;
		int prime[N/2];
		int ptr = 0;
		unsigned long counter = 0;
		prime[ptr++] = 2;
		prime[ptr++] = 3;
		for(n = 5; n <= N; n +=2){
				int flag = 0;
				for(i = 1; counter++, prime[i] * prime[i] <= n; i++){
						counter++;
						if(n % prime[i] == 0) {
								flag = 1;
								break;
						}
				}
				if(! flag)
						prime[ptr++] = n;
		}
		for(i = 0; i < ptr; i++)
				printf("%d ", prime[i]);
		printf("곱셈과 나눗셈을 실행한 횟수 : %lu\n", counter);
		printf("%d까지의 소수의 갯수는 %d 입니다.\n", N, ptr);

		return 0;
}

File: ch01/python3/ch02/prime3.py
N = 1000000

if __name__ == '__main__':
    list_prime = []
    ptr = 0
    counter = 0
    list_prime.insert(ptr, 2)
    ptr += 1
    list_prime.insert(ptr, 3)
    ptr += 1
    for n in range(5, N+1, 2):
        flag = 0
        max_index = 0
        for i in range(1, ptr):
            counter += 1
            if list_prime[i]**2 > n:
                max_index = i
                break
        for i in range(1, max_index+1):
            counter += 1
            if n % list_prime[i] == 0:
                flag = 1
                break
        if not flag:
            list_prime.insert(ptr, n)
            ptr += 1

    for i in range(0, ptr):
        print("{} ".format(list_prime[i]), end='')
    print()
    print("곱셈과 나눗셈을 실행한 횟수 : {}".format(counter))
    print("{}까지의 소수의 갯수는 {} 입니다.".format(N, ptr))
